providers/implementations/kem/hybrid_kem.c

```diff
@@ -0,0 +1,301 @@
+                                        const char **names,
+                                        uint32_t *key_lengths)
+{
+    HYBRID_PKEY_CTX *ctx =  OPENSSL_zalloc(sizeof(*ctx));
+    int i;
+
+    if (ctx == NULL)
+        return NULL;
+    ctx->num_algs = num_algs;
+    ctx->libctx = PROV_LIBCTX_OF(provctx);
+    ctx->propq = NULL;
+    ctx->names = names;
+    ctx->key_lengths = key_lengths;
+    for (i = 0; i < num_algs; i++)
+        if (0)
+            goto err;
+    return ctx;
+ err:
+    ossl_hybrid_kem_freectx(ctx);
+    return NULL;
+}
+
+void ossl_hybrid_kem_freectx(void *vctx)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    int i;
+
+    if (ctx == NULL)
+        return;
+    for (i = 0; i < ctx->num_algs; i++)
+        EVP_PKEY_CTX_free(ctx->ctxs[i]);
+    OSSL_PARAM_free(ctx->gettable);
+    OSSL_PARAM_free(ctx->settable);
+    OPENSSL_free(ctx->propq);
+    OPENSSL_free(ctx);
+}
+
+void *ossl_hybrid_kem_dupctx(void *vctx)
+{
+    HYBRID_PKEY_CTX *src = (HYBRID_PKEY_CTX *)vctx;
+    HYBRID_PKEY_CTX *dest = OPENSSL_zalloc(sizeof(*src));
+    int i;
+
+    if (dest == NULL)
+        return NULL;
+
+    dest->libctx = src->libctx;
+    dest->key_lengths = src->key_lengths;
+    dest->num_algs = src->num_algs;
+    dest->names = src->names;
+    if (src->propq != NULL
+            && (dest->propq = OPENSSL_strdup(src->propq)) == NULL)
+        goto err;
+    if (src->gettable != NULL
+            && (dest->gettable = OSSL_PARAM_dup(src->gettable)) == NULL)
+        goto err;
+    if (src->settable != NULL
+            && (dest->settable = OSSL_PARAM_dup(src->settable)) == NULL)
+        goto err;
+
+    for (i = 0; i < src->num_algs; i++)
+        if (src->ctxs[i] != NULL
+                && (dest->ctxs[i] = EVP_PKEY_CTX_dup(src->ctxs[i])) == NULL)
+            goto err;
+    return dest;
+ err:
+    ossl_hybrid_kem_freectx(dest);
+    return NULL;
+}
+
+int ossl_hybrid_kem_init(HYBRID_PKEY_CTX *ctx, const OSSL_PARAM params[],
+                         int (*initf)(EVP_PKEY_CTX *ctx,
+                                      const OSSL_PARAM params[]))
+{
+    int i;
+
+    if (!ossl_prov_is_running() || !ossl_hybrid_kem_set_ctx_params(ctx, params))
+        return 0;
+    for (i = 0; i < ctx->num_algs; i++)
+        if (!(*initf)(ctx->ctxs[i], params))
+            return 0;
+    return 1;
+}
+
+int ossl_hybrid_kem_encapsulate_init(void *vctx, void *vrecip,
+                                     const OSSL_PARAM params[])
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+
+    return ossl_hybrid_kem_init(ctx, params, &EVP_PKEY_encapsulate_init);
+}
+
+int ossl_hybrid_kem_decapsulate_init(void *vctx, void *vrecip,
+                                     const OSSL_PARAM params[])
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+
+    return ossl_hybrid_kem_init(ctx, params, &EVP_PKEY_decapsulate_init);
+}
+
+int ossl_hybrid_kem_encaps_init(HYBRID_PKEY_CTX *ctx, HYBRID_PKEY *authpriv,
```

Comment:
calling this encaps_init when you pass all the inits through it (encaps and decaps and the auth variants) seems like a poor naming choice ...